src/odemis/util/registration.py

```diff
@@ -0,0 +1,456 @@
+    """
+    Provide an initial estimate for the orientation of a square grid of points.
+
+    Parameters
+    ----------
+    xy : ndarray of shape (n, 2)
+        Array with the determined coordinates of the grid points.
+
+    Returns
+    -------
+    tform : AffineTransform
+
+    """
+    graph = nearest_neighbor_graph(xy)
+
+    # Create an array `dxy` containing the displacement vectors for all edges
+    # in the graph.
+    a, b = zip(*graph.iter_edges(False))
+    dxy = xy[b, ...] - xy[a, ...]
+
+    # The array `dxy` typically contains the lattice vectors for the 'left',
+    # 'right', 'up', and 'down' directions, but it is not guaranteed that it
+    # will contain all of them, nor that there are equal amounts of edges in
+    # each direction. The solution is to not make any distinction between
+    # 'left' and 'right', or 'up' and 'down'. This is done using a mapping of
+    # `(ρ, θ)` to `(ρ, 2*θ)`. The clustering itself is done in Cartesian
+    # coordinates to not be impacted by the branch cut at `θ = ±π`.
+    rho, theta = cartesian_to_polar(dxy)
+    dpq = polar_to_cartesian(rho, 2 * theta)
+    # minit="++" is only available in SciPy v1.2.0 and above
+    rng = check_random_state(None)
+    code_book = _kpp(dpq, 2, rng)
+    centroid, _ = scipy.cluster.vq.kmeans2(dpq, code_book, minit="matrix")
+    rho, theta = cartesian_to_polar(centroid)
+    dxy = polar_to_cartesian(rho, 0.5 * theta)
+
+    # Each rows of the array `dxy` now contains a lattice vector (direction).
+    # By transposing the array, these lattice vectors are the columns of the
+    # new matrix. This allows to directly use them as a transformation matrix.
+    # For example: let `x₁` and `x₂` be the two 2-by-1 lattice vectors, then
+    # `(x₁, x₂) * (n, m)ᵀ = n * x₁ + m * x₂`.
+    matrix = _canonical_matrix_form(numpy.transpose(dxy))
+
+    # The estimated translated is the centroid of the input coordinates.
+    translation = numpy.mean(xy, axis=0)
+
+    return AffineTransform(matrix, translation)
+
+
+def estimate_mpp_orientation(
+    xy: numpy.ndarray, shape: Tuple[int, int], transform_type: Type[T]
+) -> T:
+    """
+    Estimate the orientation of a square grid of points.
+
+    Parameters
+    ----------
+    xy : ndarray of shape (n, 2)
+        Array with the determined coordinates of the grid points.
+    shape : tuple of two ints
+        The shape of the grid given as a tuple `(height, width)`.
+    transform_type : GeometricTransform
+        The transform class to use for estimating the orientation.
+
+    Returns
+    -------
+    tform : GeometricTransform
+        The orientation of the pattern.
+
+    """
+    tform = _initial_estimate_mpp_orientation(xy)
+    grid = unit_gridpoints(shape, mode="xy")
+    correspondences = bijective_matching(tform.apply(grid), xy)
+    a, b = zip(*correspondences)
+    return transform_type.from_pointset(grid[a, ...], xy[b, ...])
+
+
+def estimate_mpp_orientation_from_img(
```

Comment:
I would rename this to `estimate_grid_orientation_from_img`

Reply:
done